### PR TITLE
[FLINK-3945] [gelly] Degree annotation for directed graphs

### PR DESCRIPTION
**Pull request by greghogan**

### CODE REVIEW

File: flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/directed/EdgeDegreesPair.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+				.setParallelism(parallelism));
+
+		// t, d(t)
+		DataSet<Vertex<K, Degrees>> vertexDegrees = input
+			.run(new VertexDegrees<K, VV, EV>()
+				.setParallelism(parallelism));
+
+		// s, t, (d(s), d(t))
+		return edgeSourceDegrees
+			.join(vertexDegrees, JoinHint.REPARTITION_HASH_SECOND)
```
Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/directed/EdgeDegreesPair.java]: I'm not sure whether it's a good idea to always provide a join hint here. It's true that the vertex set is _usually_ much smaller than the edge set, but what if that's not true? Can we at least provide a way to override this configuration?

Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/directed/EdgeDegreesPair.java]: Algorithms should not expect the user to have knowledge of the input graph or divine the best configuration. Worst case for an undirected graph is twice as many vertices as edges.

Comment (vasia) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/directed/EdgeDegreesPair.java]: I agree that we should automatically set the parameters when possible. So, I would keep this as the default setting. However, I think we should allow users that _do_ have knowledge about the input graph to either let the optimizer decide or override the default setting.

Comment (greghogan) [inline flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/asm/degree/annotate/directed/EdgeDegreesPair.java]: I'm amenable to adding this configuration if we can show noticeably improved performance. Choosing a bad `JoinHint` will result in significantly degraded performance for most graphs in addition to adding to the user's cognitive load.

**Comment by greghogan** [comment]
I'll merge this if there are no further comments.

**Pull request closed without merging**